libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationInfoProvider.kt

```diff
@@ -0,0 +1,165 @@
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.ui.maneuver.api.MapboxManeuverApi
+import com.mapbox.navigation.ui.maneuver.model.Maneuver
+import com.mapbox.navigation.ui.maneuver.model.ManeuverError
+import com.mapbox.navigation.ui.shield.model.RouteShield
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+
+/**
+ * Provides the [NavigationTemplate.NavigationInfo] and [TravelEstimate] populated with Mapbox
+ * navigation data. This can be used with a [Screen] that shows a [NavigationTemplate].
+ *
+ * @see [ActiveGuidanceScreen] for an example.
+ */
+@MapboxExperimental
+class CarNavigationInfoProvider
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+internal constructor(
+    private val services: CarNavigationInfoServices
+) : MapboxCarMapObserver {
+    /**
+     * Public constructor and the internal constructor is for unit testing.
+     */
+    constructor() : this(CarNavigationInfoServices())
+
+    private val mapUserStyleObserver = services.mapUserStyleObserver()
+    private val routeProgressObserver = RouteProgressObserver(this::onRouteProgress)
+    private val navigationObserver = mapboxNavigationForward(this::onAttached, this::onDetached)
+    private val _carNavigationInfo = MutableStateFlow(CarNavigationInfo())
+    private var currentShields = emptyList<RouteShield>()
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var carContext: CarContext? = null
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var navigationInfoMapper: CarNavigationInfoMapper? = null
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var navigationEtaMapper: CarNavigationEtaMapper? = null
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var maneuverApi: MapboxManeuverApi? = null
+
+    /**
+     * Contains data that helps populate the [NavigationTemplate] with navigation data.
+     */
+    val carNavigationInfo: StateFlow<CarNavigationInfo> = _carNavigationInfo.asStateFlow()
+
+    /**
+     * When you have a [NavigationTemplate.Builder], you can simply [apply] this lambda and it
+     * will set the navigation info and travel estimate.
+     */
+    fun setNavigationInfo(): (NavigationTemplate.Builder).() -> Unit = {
+        with(carNavigationInfo.value) {
+            navigationInfo?.let {
+                setNavigationInfo(it)
+            }
+            destinationTravelEstimate?.let {
+                setDestinationTravelEstimate(it)
+            }
+        }
+    }
```

Comment:
Why not do something like this? Looks clearer in my opinion. 
```suggestion
    fun setNavigationInfo(templateBuilder: NavigationTemplate.Builder) {
        with(carNavigationInfo.value) {
            navigationInfo?.let {
                templateBuilder.setNavigationInfo(it)
            }
            destinationTravelEstimate?.let {
                templateBuilder.setDestinationTravelEstimate(it)
            }
        }
    }
```
Usage will look like this:
```kotlin
apply { navigationInfoProvider.setNavigationInfo(this) }
```
Also lambda will be optimized with my suggestion, but not with yours.

Reply:
was looking to optimize this interface for building your own `Screen`, so when you build the NavigationTemplate it looks like this

``` kotlin
return NavigationTemplate.Builder()
...
    .apply(navigationInfoProvider.setNavigationInfo())
    .build()
```

using `fun setNavigationInfo(templateBuilder: NavigationTemplate.Builder)` works too
``` kotlin
return NavigationTemplate.Builder()
..
    .also { navigationInfoProvider.setNavigationInfo(it) }
    .build()
```

One thing i don't like about `fun setNavigationInfo(): (NavigationTemplate.Builder).() -> Unit = {` is that it does not care at all about Java

Reply:
Also like this solution to be honest, as an extension function

``` kotlin
val builder = NavigationTemplate.Builder()
...
    .setCarNavigationInfo(navigationInfoProvider)
    .build()
```

``` kotlin
fun NavigationTemplate.Builder.setCarNavigationInfo(provider: CarNavigationInfoProvider) = apply {
    with(provider.carNavigationInfo.value) {
        navigationInfo?.let {
            setNavigationInfo(it)
        }
        destinationTravelEstimate?.let {
            setDestinationTravelEstimate(it)
        }
    }
}
```

Reply:
going with your suggestion